### DISCUSSION TITLE
3주차 미팅

### DISCUSSION THREAD

**Discussion by nyj001012** [nyj001012/ft_irc#22] [General]
.

**Comment by nyj001012** [comment]
# 소켓 & 파싱
## 진행 상황
https://github.com/nyj001012/ft_irc/pull/20

# inspircd(서버) 와 irssi(클라이언트), tcpflow 사용
참고: https://80000coding.oopy.io/1ac75b59-6930-4297-9c9d-7dec31eff19d

## 사용해보며 알아낸 것
1. irssi는 `/nick yena`과 같이 `/`로 명령어임을 알리는데, 막상 tcpflow로 inspircd와 irssi간의 메시지를 보면 RFC 1459에 명시된 `NICK yena` 형태로 전달된다. 따라서 irssi에서 입력한 메시지가 자체적으로 RFC 문서에 맞게끔 1차 파싱이 되어서(`/` 떼고 `NICK yena`로 파싱) 서버로 메시지가 전달되는 것 같다.
2. 주기적으로 irssi에서 내가 굳이 입력하지 않아도 inspircd에 `PING` 명령어를 보내는 것 같다.
3. 아무리 클라이언트에서 몇 십 분이고 몇 시간이고 자리비움 상태를 해놓더라도 서버에서 클라이언트를 끊지 않는다. 예를들어 게임에서는 자리비움 상태가 된 지 30분이 지나면 강제 종료 시켜버리지만, 여기서는 그렇지 않았다.
4. 한 번의 입력에 여러 명령어를 처리하는 것은 불가능하다. (ex. `/nick yena; /msg yena hi` 또는 `/nick yena\r\n/msg yena hi`)
5. 생각보다 파싱에서 크게 신경쓸 만한 내용은 없다. 애초에 파싱이 없어도 되지 않을까 싶다. 어차피 irssi에서 RFC BNF에 맞게 파싱한 채로 서버에 메시지를 보내기 때문

## 다음 주 목표 
> collaborate with: @sejoonkimmm 

서버 예외 처리 시, 프로그램이 종료되는 부분과 종료되지 않고 계속 살아있게 하는 곳 구별
- ex) `bind()` 실패 시 error로 exit해야 맞지만, `read()`나 `write()`를 실패했다고 서버가 종료되는 건 안 됨. 현재는 throw-catch로 바로 return 1을 하며 프로그램이 종료되도록 구현되어 있음

# 의논해보면 좋을 것
- 평가를 할 때, netcat(`nc` 명령어)를 이용해서도 테스트하지만, 우리가 정한 상용 클라이언트(ex. irssi)로도 테스트를 한다. **어떤 상용 클라이언트를 사용할지 결정**해야 한다. 클라이언트마다 따르는 RFC가 다르다.
- 슬슬 IRC 기능도 구현해야 한다.
- 모르겠거나 하면 다른 카뎃분께도 SOS를 구하자.

**Reply by bartshin** [reply]
1. 현재 구현된 (구현 될) 소켓 메세지 파싱은 어떤 형태로 결과가 나올까요? 
어떤 에러 처리가 되어있고 앞으로 해야하나요?

2. 메세지 전송의 인터페이스는 어떻게 정하면 좋을까요?

**Reply by nyj001012** [reply]
1. `std::vector<t_token>`입니다. 에러 처리의 경우, BNF 형식에 맞지 않는다면 입력을 무시하는 형태로만 진행되고 별다른 에러 메시지는 출력하지 않습니다. BNF 상의 에러는 상용 클라이언트에서도 별 출력이 없더라구요
2. 메시지 전송 인터페이스는 RFC 문서대로 따라가는 게 맞을 것 같습니다. 상용 클라이언트야 명령어를 RFC에 명시된 대로 입력하지 않더라도 내부에서 RFC에 맞게 파싱한 후에 서버로 그 내용을 전달하니까요

**Reply by nyj001012** [reply]
다음 주 목표 부분은 @sejoonkimmm 님께 부탁드려도 될까요? 전반적인 서버 흐름 보기에도 좋을 것 같아서요

**Reply by sejoonkimmm** [reply]
슬랙처럼 그냥 적으면 되는 거군요...! @nyj001012 님부재이실 동안 목표 부분 제가 확인하겠습니다

**Reply by bartshin** [reply]
클라이언트 버퍼 처리하는 부분 std::array가 C++11 부터 가능합니다
우선 수정전 코드로 사용할게요 

<img width="1032" alt="Screen Shot 2023-11-13 at 8 56 54 PM" src="https://github.com/nyj001012/ft_irc/assets/65215523/c168dd2e-6e3a-4e24-935f-e7e40fbe12a5">

**Reply by sejoonkimmm** [reply]
저희... 서브젝트에 나와있는 password 부분은 일부러 아직 신경 안 쓰신 건가요?
옛날 서브젝트에서는 password가 선택이긴 한데, 지금 subject에서는 passowrd가 필수인 것 같아서요! 만약 놓치신 부분이라면 main함수 수정해놓고 password 부분 제가 만들려고 합니다

**Reply by bartshin** [reply]
password 부분도 유저 생성과 같이 구현중입니다
main 말고 다른 브랜치에 푸시 해놨어요

**Reply by sejoonkimmm** [reply]
read 실패 시 로직을 구성할 때 EAGAIN, EWOULDBLOCK 같은 c++11 errno 매크로를 사용할 수 없기 때문에 결국에는 표준 출력으로 에러를 출력하고 return buffer;를 하는 방식이 될 것 같아요 (std::runtime_error()를 throw하는 로직에서 변경)
그렇게 되면 기존에 잘못 작성하셨던 std::array를 return하게 되는데... 일단은 그대로 짜면 될까요?

**Reply by nyj001012** [reply]
> read 실패 시 로직을 구성할 때 EAGAIN, EWOULDBLOCK 같은 c++11 errno 매크로를 사용할 수 없기 때문에 결국에는 표준 출력으로 에러를 출력하고 return buffer;를 하는 방식이 될 것 같아요 (std::runtime_error()를 throw하는 로직에서 변경)
> 그렇게 되면 기존에 잘못 작성하셨던 std::array를 return하게 되는데... 일단은 그대로 짜면 될까요?

그대로 return하시면 제가 목요일에 돌아오니까 그때 array를 전부 list나 vector로 바꿀게요

**Reply by sejoonkimmm** [reply]
```c++
std::array<char, BUFFER_SIZE> Server::receiveMessage(int client_socket) {
  std::array<char, BUFFER_SIZE> buffer;  // 512바이트의 버퍼를 생성했다.

  ssize_t read_size = read(client_socket, &buffer, BUFFER_SIZE);
  if (read_size == -1)
  {
    this->closeClient(client_socket);
    std::cerr << "Error: read() failed" << std::endl;
    return buffer;
  }  // read에서 에러 발생 시 buffer 실제 irssi처럼 바로 끄지 않고 다른 로직을 생각해보기.
    
  if (read_size == 0) {
    this->closeClient(client_socket);
    return buffer;
  }
  if (this->_is_debug)
    std::cout << F_YELLOW << "[DEBUG] Message received: " << buffer.data() << FB_DEFAULT << std::endl;
  return buffer;
}
```
단순히 read에 실패했을 때 표준출력으로 에러를 뱉고, 빈 buffer를 return하는 형식이라면 이렇게만 하면 되긴 한 것 같습니다. (write 에서도 마찬가지로 read()에 실패한 client를 닫고 void return ;을 합니다)
다만, 실제 상용 IRC 서버는 errno를 사용한다는 차이 정도만 있는 것 같습니다. (jaehjoo님 출처)
만약, read()에 실패했다고 클라이언트를 종료해버리면 안 되게끔 설계를 하실 거라면, 클라이언트 종료 함수를 호출하지 않고
Server 클래스에서 임시 버퍼를 하나 생성한 뒤, 임시 버퍼에 읽기 실패한 내용을 담아 return하게 만들면 될 것 같습니다!

**Comment by bartshin** [comment]
## 디버깅 진행
https://github.com/nyj001012/ft_irc/pull/18
(웹 브라우저로 데이터 확인)

## 구현 예정
- 로우 레벨 (소켓) <-> 하이 레벨 (irc) 사이의 인터페이스
- 유저 등록
- 유저간 채팅

**Comment by sejoonkimmm** [comment]
## 서버 클래스 Exception 중 프로그램이 종료되어야 하는 부분 / 종료되면 안 되는 부분 (서버는 살아있어야 하는 부분)
---
### 종료되어야 하는 부분

1. socket()
2. bind()
3. listen() 실패

초기 설정 시스템 호출 실패의 경우, 서버 복구 불가 / 에러 메세지 출력 후 프로그램 종료해야 함

---
### 에러 처리 후 계속 실행해야 함

1. read() 
2. write()
-> 해당 클라이언트와의 연결만 종료하거나 재시도하도록 해야함. // 수정 중
3. 클라이언트에서 연결 종료
-> 서버가 해당 클라이언트와의 연결을 정리하고 서버는 계속 실행되어야 함
